text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,764 @@
+fn compile_time_default<T: const Default>() -> T {
+    const { T::default() }
+}
+```
+
+### Maybe-const trait bounds
+
+Many generic `const fn` and especially many const trait impls do not actually require a const trait impl for their generic parameters.
+As `const fn` can also be called at runtime, it would be too strict to require it to only be able to call things with const trait impls.
+Picking up the example from [the beginning](#summary):
+
+```rust
+const trait Default {
+    fn default() -> Self;
+}
+
+impl const Default for () {
+    fn default() {}
+}
+
+impl<T: Default> Default for Box<T> {
+    fn default() -> Self { Box::new(T::default()) }
+}
+
+// This function requires a `const` impl for the type passed for T,
+// even if called from a non-const context
+const fn default<T: const Default>() -> T {
+    T::default()
+}
+
+const _: () = default();
+
+fn main() {
+    let _: Box<u32> = default();
+    //~^ ERROR: <Box<u32> as Default>::default cannot be called at compile-time
+}
+```
+
+What we instead want is that, just like `const fn` can be called at runtime and compile time, we want their trait bounds' constness
+to mirror that behaviour. So we're introducing `~const Trait` bounds, which mean "const if called from const context" (slight oversimplifcation, but read on).
+
+The only thing we need to change in our above example is the `default` function, changing the `const Default` bound to a `~const Default` one.
+
+```rust
+const fn default<T: ~const Default>() -> T {
+    T::default()
+}
+```
+
+`~const` is derived from "approximately", meaning "maybe" in this context, or specifically "const impl required if called in const context".
+It is the opposite of `?` (prexisting for `?Sized` bounds), which also means "maybe", but from the other direction: `?const` (not proposed here, see the alternatives section for why it was rejected) would mean "no const impl required, even if called in const context".
+
+### `~const Destruct` trait
+
+The `Destruct` trait enables dropping types within a const context.
+
+```rust
+const fn foo<T>(t: T) {
+    // `t` is dropped here, but we don't know if we can evaluate its `Drop` impl (or that of its fields' types)
+}
+const fn baz<T: Copy>(t: T) {
+    // Fine, `Copy` implies that no `Drop` impl exists
+}
+const fn bar<T: ~const Destruct>(t: T) {
+    // Fine, we can safely invoke the destructor of `T`.
+}
+```
+
+When a value of a generic type goes out of scope, it is dropped and (if it has one) its `Drop` impl gets invoked.
+This situation seems no different from other trait bounds, except that types can be dropped without implementing `Drop`
+(as they can contain types that implement `Drop`). In that case the type's drop glue is invoked.
+
+The `Destruct` trait is a bound for whether a type has drop glue. This is trivally true for all types.
+
+`~const Destruct` trait bounds are satsifed only if the type has a `const Drop` impl or all of the types of its components
+are `~const Destruct`.
```

Comment:
This seems like a backwards compatibility hazard, similar to the existing auto traits (i.e. `Send` and `Sync`): If a struct has only `~const Destruct` fields (and no own `Drop` implementation), then adding a field with a non-const `Drop` would break backwards compatibility.

The definition of which types implement `~const Destruct` also seems wrong, I assume it should be "all components are `~const Destruct` _**and**_, if it exists, the `Drop` impl must be `const`.

I assume the `: const Destruct` (no "maybe") is also a valid bound?

Reply:
> This seems like a backwards compatibility hazard, similar to the existing auto traits (i.e. `Send` and `Sync`): If a struct has only `~const Destruct` fields (and no own `Drop` implementation), then adding a field with a non-const `Drop` would break backwards compatibility.

It already is, on stable Rust.  If we have:

```rust
pub struct S {
    x: u8,
}

// Somewhere else:

const fn f(_: S) {}
```

And then add a field:

```rust
pub struct S {
    x: u8,
    y: Box<()>,
}

// Somewhere else:

const fn f(_: S) {}
//~^ ERROR destructor of `S` cannot be evaluated at compile-time
```